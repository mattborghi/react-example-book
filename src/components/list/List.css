.List-item {
    list-style-type: none;
  }
  