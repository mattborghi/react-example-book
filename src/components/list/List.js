import { Button } from '../../components/button/Button.js';

function Item({ item, onRemoveItem }) {
  const { id, title, name } = item;
  return (
    <li className="App-item" key={id}>

      <h1>{title}</h1>
      <h2>Hello {name}!</h2>
      <Button title="Remove" onClick={() => onRemoveItem(item)} />
    </li>
  )
}

export function List({ values, onRemoveItem }) {
  return (
    <ul>
      {values.map(item => {
        return <Item key={item.id} item={item} onRemoveItem={onRemoveItem} />
      })}
    </ul>
  )
}