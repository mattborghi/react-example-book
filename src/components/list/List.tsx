import { memo } from "react";

import { Button } from "../../components/button/Button";

import { ReactComponent as Check } from "../../assets/check.svg";

export type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type ItemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
}

export type Stories = Array<Story>;

type ListProps = {
  values: Stories;
  onRemoveItem: (item: Story) => void;
}

function Item({ item, onRemoveItem } : ItemProps ) {
  const { objectID, url, title, author } = item;
  return (
    <li className="item" key={objectID}>
      <span style={{ width: "60%" }}>
        <a href={url}>{title}</a>
      </span>
      <span style={{ width: "30%" }}>{author}</span>
      <span style={{ width: "10%" }}>
        <Button onClick={() => onRemoveItem(item)}>
          <Check width="18px" height="18px" />
        </Button>
      </span>
    </li>
  );
}

export function List({ values, onRemoveItem }: ListProps) {
  return (
    <ul>
      {values.map((item) => {
        return (
          <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
        );
      })}
    </ul>
  );
}

export const MemoizedList = memo(List);
