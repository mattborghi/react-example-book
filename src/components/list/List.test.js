import renderer from 'react-test-renderer';

import { Item, List } from "./List";

describe('item', () => {
    const item = {
        objectID: 1,
        title: 'React',
        author: 'Jordan Walke',
        url: 'https://www.google.com',
        num_comments: 1,
        points: 1,
    };
    const handleRemoveItem = jest.fn();

    let component;

    beforeEach(() => {
        component = renderer.create(
            <Item item={item} onRemoveItem={handleRemoveItem} />
        );
    });

    // test inputs: item
    it('renders all properties', () => {
        expect(component.root.findByType('a').props.href).toEqual(item.url);
        expect(component.root.findAllByProps({ children: item.author }).length).toEqual(1);
        expect(component.root.findAllByProps({ children: item.title }).length).toEqual(1);
    });

    // test outputs: onRemoveItem
    it('calls onRemoveItem on button click', () => {
        component.root.findByType('button').props.onClick();
        // how many times was handleRemoveItem called?
        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
        // verify arguments passed to the function
        expect(handleRemoveItem).toHaveBeenCalledWith(item);
        // was the component rendered only once?
        expect(component.root.findAllByType(Item).length).toEqual(1);
    });

    test('renders snapshot', () => {
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('list', () => {
    const handleRemoveItem = jest.fn();
    const list = [
        {
            objectID: 12,
            title: 'React',
            author: 'Jordan Walke',
            url: 'https://www.google.com',
            num_comments: 1,
            points: 1,
        },
        {
            objectID: 2,
            title: 'Redux',
            author: 'Dan Abramov',
            url: 'https://www.react.org',
            num_comments: 10,
            points: 10,
        }
    ];
    let component;

    beforeEach(() => {
        component = renderer.create(
            <List values={list} onRemoveItem={handleRemoveItem} />
        );
    });

    it('renders two times', () => {
        expect(component.root.findAllByType(Item).length).toEqual(2);
    });

});