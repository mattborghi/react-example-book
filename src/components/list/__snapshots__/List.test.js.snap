// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item renders snapshot 1`] = `
<li
  className="item"
>
  <span
    style={
      Object {
        "width": "60%",
      }
    }
  >
    <a
      href="https://www.google.com"
    >
      React
    </a>
  </span>
  <span
    style={
      Object {
        "width": "30%",
      }
    }
  >
    Jordan Walke
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <button
      className="button button_small"
      onClick={[Function]}
    >
      <svg
        height="18px"
        width="18px"
      >
        check.svg
      </svg>
    </button>
  </span>
</li>
`;
